SignalMessaging/attachments/EditContactShareNameViewController.swift

```diff
@@ -0,0 +1,308 @@
+import Foundation
+import SignalServiceKit
+
+@objc
+public protocol ContactNameFieldViewDelegate: class {
+    func nameFieldDidChange()
+}
+
+// MARK: -
+
+class ContactNameFieldView: UIView {
+    weak var delegate: ContactNameFieldViewDelegate?
+
+    let name: String
+    let initialValue: String?
+
+    var valueView: UITextField!
+
+    var hasUnsavedChanges = false
+
+    // MARK: - Initializers
+
+    @available(*, unavailable, message: "use other constructor instead.")
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("Unimplemented")
+    }
+
+    required init(name: String, value: String?, delegate: ContactNameFieldViewDelegate) {
+        self.name = name
+        self.initialValue = value
+        self.delegate = delegate
+
+        super.init(frame: CGRect.zero)
+
+        self.isUserInteractionEnabled = true
+        self.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(wasTapped)))
+
+        createContents()
+    }
+
+    func createContents() {
+        let vMargin = CGFloat(10)
+        self.layoutMargins = UIEdgeInsets(top: vMargin, left: 0, bottom: vMargin, right: 0)
+
+        let stackView = UIStackView()
+        stackView.axis = .horizontal
+        stackView.alignment = .center
+        stackView.layoutMargins = .zero
+        stackView.spacing = 10
+        self.addSubview(stackView)
+        stackView.autoPinTopToSuperviewMargin()
+        stackView.autoPinBottomToSuperviewMargin()
+        stackView.autoPinLeadingToSuperviewMargin()
+        stackView.autoPinTrailingToSuperviewMargin()
+
+        let nameLabel = UILabel()
+        nameLabel.text = name
+        nameLabel.font = UIFont.ows_dynamicTypeBody
+        nameLabel.textColor = UIColor.ows_materialBlue
+        nameLabel.lineBreakMode = .byTruncatingTail
+        stackView.addArrangedSubview(nameLabel)
+        nameLabel.setContentHuggingHigh()
+        nameLabel.setCompressionResistanceHigh()
+
+        valueView = UITextField()
+        if let initialValue = initialValue {
+            valueView.text = initialValue
+        }
+        valueView.font = UIFont.ows_dynamicTypeBody
+        valueView.textColor = UIColor.black
+        stackView.addArrangedSubview(valueView)
+
+        valueView.addTarget(self, action: #selector(textFieldDidChange), for: .editingChanged)
+    }
+
+    func wasTapped(sender: UIGestureRecognizer) {
+        Logger.info("\(self.logTag) \(#function)")
+
+        guard sender.state == .recognized else {
+            return
+        }
+
+        valueView.becomeFirstResponder()
+    }
+
+    func textFieldDidChange(sender: UITextField) {
+        Logger.info("\(self.logTag) \(#function)")
+
+        hasUnsavedChanges = true
+
+        guard let delegate = self.delegate else {
+            owsFail("\(logTag) missing delegate.")
+            return
+        }
+
+        delegate.nameFieldDidChange()
+    }
+
+    public func value() -> String {
+        guard let value = valueView.text else {
```

Comment:
Generally I'm all about returning non-null values, but we sometimes use a nil `String?` to indicate a value is *missing*, e.g. if someone doesn't have a name prefix.

Do we make other code simpler by returning an empty string vs nil here?

I don't want to confuse empty strings as real values elsewhere - the only thing worse than an unnecessary nil check is an unnecessary `length == 0` check.

Take it or leave it though, maybe we can't avoid those length checks anyway.

Reply:
I think this is already fixed in a subsequent PR.